infor-design/enterprise
# ng#705 - Implement Modal Manager

**Explain the _details_ for making this change. What existing problem does the pull request solve?**
TBD

**Related github/jira issue (required)**:
Closes infor-design/enterprise-ng#705
Related to #2287
Related to #3633
Related to #3682

**Steps necessary to review your pull request (required)**:
Pull this branch, build, and run the demoapp.  Also pull [`ids-enterprise-ng`](https://github.com/infor-design/enterprise-ng) master branch, install, and either symlink or `npm link` this branch of `ids-enterprise` to that project to test the original issue.

_Test the original NG issue:_
- Build/Serve NG and open http://localhost:4200/ids-enterprise-ng-demo/#/ids-enterprise-ng-demo/modal-dialog
- Click the button labeled "Nested". The Modal should appear.
- Click the "Show another Modal" button inside the Modal.  The modal should appear to close and open another modal on top.  There should not be a double overlay or overlapping modal windows.
- Click that button a few more times to open a few more modals deep.
- Start pressing the ESCAPE key individually.  The modals should close one-by-one.  Holding the ESCAPE key down should close each one relatively quickly without throwing errors.

_EP Test for closing the escape key in nested Modals:_
- Run EP Demoapp and open http://localhost:4000/components/modal/test-nested.html
- Click the "Show Nested Modal" button
- When the modal opens, click "Show another modal".  Keep clicking the buttons inside modals that open more modals until it's no longer possible, and you get to "Lowest Modal in the Nest".  When clicking through, there should no longer be multiple overlays, and only one modal should be present on screen at a time.
- Start pressing the Escape key until all the modals are closed.  Eventually no more should be present, and the overlay should hide.

_Focus Test 1:_
- Run EP Demoapp and open http://localhost:4000/components/modal/test-external-components.html
- Open the modal.  When it opens, keyboard focus should be present on the input field labeled "Standard Input"
- Press the "Tab" key once.  The Dropdown should become focused.
- Press the Down arrow to open the Dropdown.  Press the up/down arrows.  Keyboard focus should be possible within the Dropdown list without it closing.  
- Press the Tab key to close the Dropdown list and focus on the Menu Button.  Press "Enter/Return" to open the Popupmenu.
- Press the Escape key to close the Popupmenu.  The Modal should remain open.
- Press the Tab Key to focus the Popover trigger button.  Press the Enter/Return key to open the Popover.
- Press the Escape key to close the Popover.  The Modal should remain open.
- Press the Escape key again to close the Modal.  The Modal's trigger button should become focused when the Modal closes.
- *NOTE:* This one is covered by a new e2e test.

_Focus Test 2:_
- Run the EP Demoapp and open http://localhost:4000/components/lookup/test-multiple-custom-modals.html
- Click the "Select State" button.  When the modal opens, the "Date Field" datepicker should be focused.
- Click the trigger button in the "Products" lookup.  The Lookup should open, and the "More Actions" button should be focused.
- Press the Enter Key to open the More Actions menu.
- Press the Escape Key.  The More Actions Menu should close, but the Lookup should remain open.
- Click on any row from the Lookup.  The Lookup should close, and focus should return to its corresponding Input field in the first Modal, now populated with the row number.

_Focus Test 3:_
- Run the EP Demoapp and open http://localhost:4000/components/modal/test-no-autofocus.html
- Click "Show Modal (no focus)" to open the modal.  When the modal opens, no field should be focused.
- Press the Escape key.  The modal should close regardless of whether an element inside is focused or not, if the Modal is open and no other inner-components are currently in control of the Escape key.

**Included in this Pull Request**:
